azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,175 @@
+
+    @staticmethod
+    def is_policy_enforcement_enabled():
+        """Check whether user has opted into policy enforcement feature"""
+        # TODO: The conf flag will be removed post private preview. Before public preview, add checks
+        # according to the planned user experience (TBD).
+        return conf.get_extension_policy_enabled()
+
+    @staticmethod
+    def check_policy_enforcement_supported():
+        """
+        Check that both platform architecture and distro/version are supported.
+        If supported, do nothing.
+        If not supported, raise PolicyError with user-friendly error message.
+        """
+        osutil = get_osutil()
+        arch = osutil.get_vm_arch()
+        # TODO: surface as a user error with clear instructions for fixing
+        msg = "Attempted to enable policy enforcement, but feature is not supported on "
+        if arch not in POLICY_SUPPORTED_ARCHITECTURE:
+            msg += " architecture " + str(arch)
+        elif DISTRO_NAME not in POLICY_SUPPORTED_DISTROS_MIN_VERSIONS:
+            msg += " distro " + str(DISTRO_NAME)
+        else:
+            min_version = POLICY_SUPPORTED_DISTROS_MIN_VERSIONS.get(DISTRO_NAME)
+            if DISTRO_VERSION < min_version:
+                msg += " distro " + DISTRO_NAME + " " + DISTRO_VERSION + ". Policy is only supported on version " + \
+                        str(min_version) + " and above."
+            else:
+                return  # do nothing if platform is supported
+        raise PolicyError(msg)
+
+
+    @property
+    def policy_engine_enabled(self):
+        """This property tracks whether the feature is enabled and Regorus engine has been successfully initialized"""
+        return self._policy_engine_enabled
+
+    def evaluate_query(self, input_to_check, query):
+        """
+        Input_to_check is the input we want to check against the policy engine (ex: extensions we want to install).
+        Input_to_check should be a dict. Expected format:
+        {
+            "extensions": {
+                "<extension_name_1>": {
+                    "signingInfo": {
+                        "extensionSigned": <true, false>
+                    }
+                }, ...
+        }
+
+        The query parameter specifies the value we want to retrieve from the policy engine.
+        Example format for query: "data.agent_extension_policy.extensions_to_download"
+        """
+        # This method should never be called if policy is not enabled, this would be a developer error.
+        if not self.policy_engine_enabled:
+            raise PolicyError("Policy enforcement is disabled, cannot evaluate query.")
+
+        try:
+            full_result = self._engine.eval_query(input_to_check, query)
+            debug_info = "Rule file is located at '{0}'. \nFull query output: {1}".format(self._engine.rule_file, full_result)
+            if not full_result.get('result') or not isinstance(full_result['result'], list):
```

Comment:
`not full_result.get('result')`
Are we checking for None here? Looks like 'result' value type is list so we should probably do explicit check for None here instead of 'not full_result...'
```suggestion
            if full_result.get('result') is None or not isinstance(full_result['result'], list):
```

Reply:
+1

@mgunnala We avoid those implicit comparisons. We check "if foo/if not foo" only for actual bools. Comparisons against None, empty list, empty string, etc, should be explicit

Reply:
Added explicit None checks for full_result, result, expression and value